core/src/main/java/org/jdbi/v3/core/ConfigOnlyExtension.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core;
+
+import org.jdbi.v3.core.extension.ExtensionConfig;
+import org.jdbi.v3.core.extension.ExtensionFactory;
+
+class ConfigOnlyExtension<C extends ExtensionConfig<C>> implements ExtensionFactory<C>
```

Comment:
To elaborate on this comment and others:

There is nothing extension-specific about `ExtensionConfig`. We introduced it because extensions needed _somewhere_ to keep configuration. But if you think about it, we already maintain config objects all over the place: the registries. What if we promoted configuration objects to first-class citizens?

Proposal:

* Rename existing `JdbiConfig` to `ConfigRegistry`. Make it public.
* Rename `ExtensionConfig` to `JdbiConfig`.
* Modify the argument registry, mapping registry, collector registry, and extension registry to implement `JdbiConfig`. It might be valuable to make their `createChild()` method implementations use a chain-of-responsibility pattern to minimize copied objects at runtime.
* Fold the statement rewriter, timing collector, and statement attributes into a new config type. `SqlStatementConfig` perhaps?
* Refactor `ConfigRegistry` so it is just a registry of `JdbiConfig` types, similar to the other registries. A `JdbiConfig` of a given type is instantiated automatically the first time it is accessed.
* Replace `ConfigRegistry.copyOf` with `ConfigRegistry.createChild` for uniformity with `JdbiConfig`. Possibly also make `ConfigRegistry` use chain of responsibility, to defer copying parent config objects until needed.
* Refactor extensions so that they get the `ConfigRegistry`, not just not just the one config type. To that end, remove the `C extends ExtensionConfig<C>` generic parameter, and the `createConfig` method from `ExtensionFactory`.
* Add accessor methods to `Jdbi`, `Handle`, `SqlStatement`, and `StatementContext`
* Refactor SQL Object so that `Handler` has access to the entire config registry, not just `SqlObjectConfig`.

The cool thing about this is that anybody can add their own new config concepts, and then everything can access that configuration with the same ease as the built-in config like the mapping registry.

@stevenschlansker @arteam @brianm What do you think?